service/matching/matching_engine.go

```diff
@@ -2928,3 +2937,125 @@ func largerBacklogAge(rootBacklogAge *durationpb.Duration, currentPartitionAge *
 	}
 	return currentPartitionAge
 }
+
+func buildRateLimitConfig(update *workflowservice.UpdateTaskQueueConfigRequest_RateLimitUpdate, updateTime *timestamppb.Timestamp) *taskqueuepb.RateLimitConfig {
+	var rateLimit *taskqueuepb.RateLimit
+	if r := update.GetRateLimit(); r != nil {
+		rateLimit = &taskqueuepb.RateLimit{RequestsPerSecond: r.RequestsPerSecond}
+	}
+	return &taskqueuepb.RateLimitConfig{
+		RateLimit: rateLimit,
+		Metadata: &taskqueuepb.ConfigMetadata{
+			Reason:     update.GetReason(),
+			UpdateTime: updateTime,
+		},
+	}
+}
+
+func prepareTaskQueueUserData(
+	tqud *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *persistencespb.TaskQueueUserData {
+	data := common.CloneProto(tqud)
+	if data == nil {
+		data = &persistencespb.TaskQueueUserData{}
+	}
+	if data.PerType == nil {
+		data.PerType = make(map[int32]*persistencespb.TaskQueueTypeUserData)
+	}
+	tqType := int32(taskQueueType)
+	if data.PerType[tqType] == nil {
+		data.PerType[tqType] = &persistencespb.TaskQueueTypeUserData{}
+	}
+	if data.PerType[tqType].Config == nil {
+		data.PerType[tqType].Config = &taskqueuepb.TaskQueueConfig{}
+	}
+	return data
+}
+
+func constructUpdateTaskQueueConfigResponse(
+	taskQueueUserData *persistencespb.TaskQueueUserData,
+	taskQueueType enumspb.TaskQueueType,
+) *matchingservice.UpdateTaskQueueConfigResponse {
+	if taskQueueUserData.GetPerType() == nil {
+		return &matchingservice.UpdateTaskQueueConfigResponse{}
+	}
```

Comment:
you can leave out this part, looking up something in a nil map returns "zero, false" so the next case handles it